helix-view/src/clipboard.rs

```diff
@@ -6,7 +6,7 @@ use std::borrow::Cow;
 pub trait ClipboardProvider: std::fmt::Debug {
     fn name(&self) -> Cow<str>;
     fn get_contents(&self) -> Result<String>;
-    fn set_contents(&self, contents: String) -> Result<()>;
+    fn set_contents(&mut self, contents: String) -> Result<()>;
```

Comment:
I wonder if the implementation can be refactored to prevent code duplication; for example `paste_primary_clipboard_impl` and `paste_clipboard_impl` differ by only two tokens. My suggestion would be to handle the distinction between primary and normal clipboard in `ClipboardProvider` itself:

```rust
pub enum ClipboardType {
	Normal,
	Primary,
}

pub trait ClipboardProvider: std::fmt::Debug {
    fn name(&self) -> Cow<str>;
    fn get_contents(&self, cliptype: ClipboardType) -> Result<String>;
    fn set_contents(&mut self, contents: String, cliptype: ClipboardType) -> Result<()>;
}
```

Reply:
Well, this will limit the possible scenarios or need further refactoring. Now you can have `win32yank` for clipboard and `none` for primary. Can be refactored this way: every provider will redefine in-memory storage. Cons: I found out using xsel is better than wl-copy for selection, since selection is modified frequently.

Reply:
> every provider will redefine in-memory storage.

I'm not quite sure I follow what you meant here, but as for the drawback you mentioned, can't it be a nested conditional inside the check for wayland like so:

```rust
    // ...
    } else if env_var_is_set("WAYLAND_DISPLAY") && exists("wl-copy") && exists("wl-paste") {
        let mut provider = command_provider! {
            paste => "wl-paste", "--no-newline";
            copy => "wl-copy", "--type", "text/plain";
        };
        if exists("xsel") {
            // None by default to account for windows and mac
            provider.get_primary_cmd = Some("xsel cmd");
            provider.set_primary_cmd = Some("xsel cmd");
        }
        provider
    } else if env_var_is_set("DISPLAY") && exists("xclip") {
    // ...
```

Reply:
Yes, looks great, thanks for advice!

Reply:
I will do this way: create `none` initially, and then add commands by if-else chains. On windows create windows native provider and implement fallback primary clipboard there (Ctrl+C clipboard is handled by winapi wrapper, selection stored inside a variable)